lib/Sema/TypeCheckCircularity.cpp

```diff
@@ -601,3 +607,43 @@ bool CircularityChecker::diagnoseInfiniteRecursion(CanType parentType,
 
   return true;
 }
+
+// Show a warning if the enum is inconstructible. The outcome of this method
+// is irrelevant.
+void CircularityChecker::diagnoseInconstructible(EnumDecl *E) {
+
+  auto containsType = [](TupleType *tuple, Type E) -> bool {
+    for (auto type: tuple->getElementTypes()) {
+      if (type->isEqual(E))
+        return true;
+    }
+    return false;
+  };
+  auto isInconstructible = [containsType, E]() -> bool {
+    auto elts = E->getAllElements();
+    if (elts.empty())
+      return false;
+
+    for (auto elt: elts) {
+      if (!elt->hasInterfaceType() ||
+          !(elt->isIndirect() || E->isIndirect()))
+        return false;
+
+      auto argTy = elt->getArgumentInterfaceType();
```

Comment:
You can simplify this and handle nested tuples by using `TypeBase::isStructurallyUninhabited`

Reply:
This is a bit unexpected for me, can you explain why a tuple containing an uninhabited type is uninhabited? I also don't understand how this can help to simplify.

Reply:
> This is a bit unexpected for me, can you explain why a tuple containing an uninhabited type is uninhabited?

```swift
enum Foo {
  case bar(Never)
}
```

Reply:
This is misleading. If I understand correctly now, uninhabited means 'not constructible'. So you can't construct a tuple that has an empty enum, because you can't instantiate an empty enum. The first thing that comes to mind when I see uninhabited is a type with no explicit members. Maybe implicit as well. Well, actually, if a type has neither explicit nor implicit members, it is inconstructible. But it's still a somewhat confusing naming. 

Do I correctly understand that you want me to additionally check (rather than simplify) if an argument is uninhabited so that such cases are also counted in favor of inconstructability? The diagnostic would then be misleading the way it is now. 

It can be changed then to `enum is not constructible because all cases are either recursive or contain a non-constructible type`.

Reply:
But, I'm not sure if it's a good idea to warn if there are 'uninhabited' cases. Similar to how we don't warn about empty enums

Reply:
In other words, using the terminology Slava gave, I don't think we should warn about uninhabited types (The issue faces non-well-formed types, or unconstructibly infinite types – the way @jckarter describes them in the report).
```swift
enum Foo1 { case A(Never), case B(Foo1) } // uninhabited

enum Foo2 { case A(Int, Foo2), case B(Foo1) // non-well-formed
```